torchtune/data/metrics/_metric_transform.py

```diff
@@ -0,0 +1,133 @@
+
+        Args:
+            sample (dict[str, Any]): The sample dictionary to generate metrics from
+
+        Returns:
+            list[Metric]: List of metrics generated for this sample
+
+        Raises:
+            NotImplementedError: If subclass does not implement this method.
+        """
+        raise NotImplementedError("Subclasses must implement _generate_metrics method")
+
+    def __call__(self, sample: dict[str, Any]) -> dict[str, Any]:
+        """Apply transform to sample, adding generated metrics."""
+        if self.dataset_name is None or self.new_metric is None:
+            raise RuntimeError(
+                "set_dataset_name() must be called before using the transform."
+            )
+
+        # Generate metrics for this sample
+        metrics = self._generate_metrics(sample)
+
+        # Add to existing metrics list or create new one
+        if "metrics" not in sample:
+            sample["metrics"] = []
+        sample["metrics"].extend(metrics)
+        return sample
+
+
+class DefaultTrainingMetricTransform(MetricTransform):
+    """Generates training metrics: samples_seen, tokens_seen, seq_len distribution.
+
+    For details about MetricTransform base class behavior, see the parent class docstring.
+
+    Tracked metrics:
+    - samples_seen: Cumulative count of samples processed (SUM aggregation)
+    - tokens_seen: Cumulative sum of all tokens processed (SUM aggregation)
+    - seq_len: Distribution of sequence lengths (DISTRIBUTION aggregation)
+
+    Example:
+        >>> transform = DefaultTrainingMetricTransform()
+        >>> transform.set_dataset_name("alpaca")
+        >>>
+        >>> sample = {"tokens": [1, 2, 3, 4, 5]}  # 5 tokens
+        >>> metrics = transform._generate_metrics(sample)
+        >>> # Creates:
+        >>> # [
+        >>> #   Metric(dataset_name="alpaca", name="samples_seen", value=1, agg_type=AggregationType.SUM),
+        >>> #   Metric(dataset_name="alpaca", name="tokens_seen", value=5, agg_type=AggregationType.SUM),
+        >>> #   Metric(dataset_name="alpaca", name="seq_len", value=5, agg_type=AggregationType.DISTRIBUTION)
+        >>> # ]
+    """
+
+    def _generate_metrics(self, sample: dict[str, Any]) -> list[Metric]:
+        if self.new_metric is None:
+            raise RuntimeError(
+                "set_dataset_name() must be called before using the transform."
+            )
+
+        # Determine token key
+        token_key = "tokens" if "tokens" in sample else "input_ids"
+        token_len = len(sample.get(token_key, []))
+
+        # Create metrics for this sample
+        return [
+            self.new_metric(name="samples_seen", value=1, agg_type=AggregationType.SUM),
+            self.new_metric(
+                name="tokens_seen", value=token_len, agg_type=AggregationType.SUM
+            ),
+            self.new_metric(
+                name="seq_len", value=token_len, agg_type=AggregationType.DISTRIBUTION
+            ),
```

Comment:
A minor thing, but to me metrics having the same value but different aggregation types should not actually be represented as distinct metrics. Like I should be able to just define how a metric is computed for a given sample, then separately choose different types of aggregation as needed